prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1408 @@
+
+Like all REQ packets, a LONGREAD_RTM consists of 3 parts: mandatoary header, REQ optional
+header and the application data. However, the application data part of a LONGREAD_RTM is
+special: it is not the data in the application's send buffer, but information of the
+sender buffer.
+
+In long-read message sub-protocol, sender need to construct an `read_iov`, which is
+an array of `efa_rma_iov` of application's send buffer. The `read_iov` is used
+as the appliction data in the LONGREAD_RTM packet.
+
+The binary format of a LONGREAD_RTM packet's mandatory header is listed in table 4.1
+
+Table: 4.1 the binary format of a LONGREAD_RTM packet's mandatory header
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `msg_id`         | 4 | integer | `uint32_t` | message ID |
+| `msg_length`     | 4 | integer | `uint64_t` | total length of the message |
+| `send_id`        | 4 | integer | `uint32_t` | ID of the receive operation  |
+| `read_iov_count` | 4 | integer | `uint32_t` | number of iov to read |
+
+Noticing the new field `read_iov_count`, which is number of `struct efa_rma_iov` in `read_iov`.
+
+To construct `read_iov`, sender need to make sure the send buffer is registered with EFA device and fill
+the registration key in `read_iov`.
+
+There are two ways to achieve that:
+
+First, if the buffer has already been registered with device, application will provide
+a memory descriptor along with the sender buffer, registration key can be extracted from the descriptor;
+
+Second, if the buffer has not been registered with EFA device, sender need to register the buffer,
+and can get the key from the registration. Note because memory registration is a limited resource,
+it is possible for memory registration to fail and sender need to be able to handle the case.
+
+Upon receiving a long-read RTM, the receiver will use RDMA read to copy data from application's
+send buffer to application's receive buffer (avoiding copy). That is why this protocol is
+sometime referred as zero-copy.
+
+After all read is finished, the receiver will send an EOR packet to the sender to notify it
+the work is done.
+
+The binary format of the EOR packet is listed in table 4.2
+
+Table: 4.2 the format of an EOR packet
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `send_id`        | 4 | integer | `uint32_t` | ID of the send operation |
+| `recv_id`        | 4 | integer | `uint32_t` | ID of the receive operation |
+| `connid`         | 4 | integer | `uint32_t` | sender connection ID, optional |
```

Comment:
The specific definition of REQ packets (which can stack header items per 3.1) needs to be more clearly defined. What specific packets are we calling REQ, vs which we are not. It ends up being odd that we are specifying connid specifically in some cases and assumed in others.

Reply:
As suggested, in table 1.2 when we introduce all packet types, I added a column to indicate which packet type is REQ, which is not.